encoding/dot/dot.go

```diff
@@ -0,0 +1,370 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package dot implements GraphViz DOT marshaling of graphs.
+package dot
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/gonum/graph"
+)
+
+// Node is a DOT graph node.
+type Node interface {
+	// DOTID returns a DOT node ID.
+	//
+	// An ID is one of the following:
+	//
+	//  - a string of alphabetic ([a-zA-Z\x80-\xff]) characters, underscores ('_') or
+	//    digits ([0-9]), not beginning with a digit
+	//  - a numeral [-]?(.[0-9]+ | [0-9]+(.[0-9]*)?)
+	//  - a double-quoted string ("...") possibly containing escaped quotes (\");
+	//  - an HTML string (<...>).
+	DOTID() string
+}
+
+// Attributers are graph.Graph values that specify top-level DOT
+// attributes.
+type Attributers interface {
+	DOTAttributers() (graph, node, edge Attributer)
+}
+
+// Attributer defines graph.Node or graph.Edge values that can
+// specify DOT attributes.
+type Attributer interface {
+	DOTAttributes() []Attribute
+}
+
+// Attribute is a DOT language key value attribute pair.
+type Attribute struct {
+	Key, Value string
+}
+
+// Porter defines the behavior of graph.Edge values that can specify
+// connection ports for their end points.
+type Porter interface {
+	FromPort() (port, compass string)
```

Comment:
Assumedly the user of this package will (have to) be familiar with the DOT format, but still, shouldn't there be something written about compass values here or about strict argument to Marshal() below? Perhaps also include the link to the DOT grammar somewhere?